src/main/java/ch/njol/skript/Skript.java

```diff
@@ -749,29 +715,51 @@ protected void afterErrors() {
 								} catch (InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException | NoSuchMethodException | SecurityException e) {
 									Skript.exception(e, "Failed to initalize test JUnit classes.");
 								}
-								if (ignored > 0)
+								if (ignored.get() > 0)
 									Skript.warning("There were " + ignored + " ignored test cases! This can mean they are not properly setup in order in that class!");
 
-								info("Completed " + tests + " JUnit tests in " + size + " classes with " + fails + " failures in " + milliseconds + " milliseconds.");
 							}
 						}
-						double display = shutdownDelay / 20;
-						info("Testing done, shutting down the server in " + display + " second" + (display <= 1D ? "" : "s") + "...");
-						// Delay server shutdown to stop the server from crashing because the current tick takes a long time due to all the tests
-						Bukkit.getScheduler().runTaskLater(Skript.this, () -> {
-							if (TestMode.JUNIT && !EffObjectives.isJUnitComplete())
-								EffObjectives.fail();
-
-							info("Collecting results to " + TestMode.RESULTS_FILE);
-							String results = new Gson().toJson(TestTracker.collectResults());
-							try {
-								Files.write(TestMode.RESULTS_FILE, results.getBytes(StandardCharsets.UTF_8));
-							} catch (IOException e) {
-								Skript.exception(e, "Failed to write test results.");
+
+						CompletableFuture.runAsync(() -> {
+							if (TestMode.JUNIT) {
+								info("Running async JUnit tests...");
+								try {
+									for (Class<?> clazz : asyncTests) {
+										runTest(clazz, shutdownDelay, tests, milliseconds, ignored, fails);
+									}
+								} catch (InstantiationException | IllegalAccessException | InvocationTargetException |
+										 NoSuchMethodException e) {
+									Skript.exception(e, "Failed to initalize test JUnit classes.");
```

Comment:
Skript.exception collects a bunch of data about the server and Bukkit. I don't believe it can be called async.